File: vane-portals/src/main/java/org/oddlama/vane/portals/event/PortalLinkConsoleEvent.java
package org.oddlama.vane.portals.event;

import java.util.List;
import org.bukkit.block.Block;
import org.bukkit.entity.Player;
import org.bukkit.event.HandlerList;
import org.jetbrains.annotations.Nullable;
import org.oddlama.vane.portals.portal.Portal;

public class PortalLinkConsoleEvent extends PortalEvent {

    private static final HandlerList handlers = new HandlerList();
    private Player player;
    private Portal portal;
    private Block console;
    private List<Block> portalBlocks;
    private boolean checkOnly;
    private boolean cancelIfNotOwner = true;

    public PortalLinkConsoleEvent(
        final Player player,
        final Block console,
        final List<Block> portalBlocks,
        boolean checkOnly,
        @Nullable final Portal portal
    ) {
        this.player = player;
        this.console = console;
        this.portalBlocks = portalBlocks;
        this.checkOnly = checkOnly;
        this.portal = portal;
    }

    public void setCancelIfNotOwner(boolean cancelIfNotOwner) {
        this.cancelIfNotOwner = cancelIfNotOwner;
    }

    public Player getPlayer() {
        return player;
    }

    public Block getConsole() {
        return console;
    }

    public List<Block> getPortalBlocks() {
        return portalBlocks;
    }

    public boolean checkOnly() {
        return checkOnly;
    }

    public @Nullable Portal getPortal() {
        return portal;
    }

    public HandlerList getHandlers() {
        return handlers;
    }

    public static HandlerList getHandlerList() {
        return handlers;
    }

    @Override
    public boolean isCancelled() {
        var cancelled = super.isCancelled();
        if (cancelIfNotOwner && portal != null) {
            cancelled |= !player.getUniqueId().equals(portal.owner());
        }
        return cancelled;
    }
}

File: vane-portals/src/main/kotlin/org/oddlama/vane/portals/event/PortalLinkConsoleEvent.kt
package org.oddlama.vane.portals.event

import org.bukkit.block.Block
import org.bukkit.entity.Player
import org.bukkit.event.HandlerList
import org.oddlama.vane.portals.portal.Portal

/**
 * Fired while validating or linking a console to a portal.
 *
 * @property player the acting player.
 * @property console the console block to link.
 * @property portalBlocks blocks that will become part of the portal.
 * @property checkOnly whether this is a dry-run validation.
 * @property portal existing portal when relinking, or null when creating.
 */
class PortalLinkConsoleEvent(
    @JvmField val player: Player,
    @JvmField val console: Block?,
    val portalBlocks: MutableList<Block?>?,
    private val checkOnly: Boolean,
    @JvmField val portal: Portal?
) : PortalEvent() {
    /** Controls whether non-owners are automatically denied. */
    private var cancelIfNotOwner = true

    /** Sets whether linking is automatically cancelled for non-owners. */
    fun setCancelIfNotOwner(cancelIfNotOwner: Boolean) {
        this.cancelIfNotOwner = cancelIfNotOwner
    }

    /** Returns whether this event is in validation-only mode. */
    fun checkOnly() = checkOnly

    /** Returns Bukkit handlers for this event. */
    override fun getHandlers() = handlerList

    /** Returns true when explicitly cancelled or blocked by ownership checks. */
    override fun isCancelled() = super.isCancelled() || (cancelIfNotOwner && portal != null && player.uniqueId != portal.owner())

    /** Static Bukkit handler list for this event type. */
    companion object {
        /** Shared handler list used by Bukkit's event system. */
        @JvmStatic val handlerList: HandlerList = HandlerList()
    }
}
